### PR TITLE
Register SSZ snappy protocol messages

### PR DESCRIPTION
**Pull request by b-m-f**
## Issue Addressed

https://github.com/sigp/lighthouse/issues/581

## Proposed Changes

- Create new SSZSnappy codec
- Register new codec with higher precedence than the normal SSZ
- Implement new codec for outbound and inbound traffic

## Additional Info
- 
-

### CODE REVIEW

File: beacon_node/eth2-libp2p/src/rpc/protocol.rs [hunk 1]
```diff
@@ -168,12 +190,18 @@ impl RPCRequest {
     pub fn supported_protocols(&self) -> Vec<ProtocolId> {
         match self {
             // add more protocols when versions/encodings are supported
+            RPCRequest::Hello(_) => vec![ProtocolId::new("hello", "1", "ssz_snappy")],
```
Comment (AgeManning) [inline beacon_node/eth2-libp2p/src/rpc/protocol.rs]: This function is used on line 184 above. Basically for each request we receive we need to tell libp2p which protocols we support. Here these are stored as a list, and we should add protocols to the list for each request. 

Currently, this match would only ever return the `ssz_snappy` protocol id and cannot return the `ssz` anymore.

**Comment by b-m-f** [comment]
Heyho, sorry for the absence. Another project was pretty demanding. But I hope to finish off the PR this week :)

**Comment by b-m-f** [comment]
Hi @AgeManning,

For a simple idea of the snappy ssz codec I have replace the `UviBytes` with the [snap](https://github.com/BurntSushi/rust-snappy) library directly. 
Is this too naive and do the `UviBytes` need to be kept in Response and Request?

I was also wondering about how to test this. 
Heres an idea:
Compare the output of, lets say a Goodbye request that leaves the system, with a hardcoded value of how this request would look like when it is snappy compressed via the library directly.

This will of course put trust in the library, but that seems fine to me.

**Comment by AgeManning** [comment]
Hey @b-m-f 
Sorry for the late reply. I've been very distracted in other areas. 

Replacing the `UviBytes` codec with the snap library is a good solution. I tried to make this part modular so that we could do such things. However we still have a bit of code duplication, but I guess this is fine for now (until we can think of a better way of combining these codecs). 

Also, we need to be careful of decompression bombs. So before decompressing, we should use the `snap::decompress_ken` to ensure the decompressed size is less than the `max_packet_size` before attempting to decompress it. 

Your testing idea sounds great to me. I have some RPC tests: https://github.com/sigp/lighthouse/blob/master/beacon_node/eth2-libp2p/tests/rpc_tests.rs 

If you prioritize the snappy compression in the protocols, these RPC tests will then test the ssz_compression also.

**Comment by pawanjay176** [comment]
Hey @b-m-f  Are you still working on this issue? We wanted to get the compression done for the next release. If you want to push what you have so far, I'd be happy to finish it off with the testing :)
Let me know. Cheers!

**Comment by b-m-f** [comment]
Hi @pawanjay176,

I wont have time to work on this before next Monday. Could probably have it done by Wednesday next week.
The pushed changes are all that I have so far.

If you like feel free to continue on this!

**Comment by b-m-f** [comment]
Hi @pawanjay176,

**Comment by pawanjay176** [comment]
Hey @b-m-f , sorry got caught up with some other stuff. Haven't worked on this PR. I don't think we'll be targeting this for the next release so its not urgent as I said earlier! Maybe the release after that with other networking improvements.
Please feel free to continue working on it if you want :) Sorry for the confusion!

**Comment by b-m-f** [comment]
@pawanjay176 okay. Thanks for the update!

I will try to push the tests soon and then this PR should be done I hope. I am quite busy these days as well, just moved and have big holidays coming up.

**Comment by b-m-f** [comment]
Hi @pawanjay176,

just saw that you self-assigned this. Thank you again for taking over and I hope you can salvage something of what I have commited :)
When I have less things going on I hope to contribute to this codebase again.

**Comment by protolambda** [comment]
I started some discussion in the [networking channel in the eth2 discord](https://discord.gg/vsghbz4) about length prefixes of snappy-compressed ssz payloads in sync rpc. Appreciate if you can review what works best, and share what this PR is headed to with spec interpretation.

**Comment by AgeManning** [comment]
Hey @b-m-f 
Thanks for all the work on this!

I've made some significant changes to the overall networking layout. I'll be merging these down in the coming weeks and will likely need to integrate this into the new format. 

If you don't mind. Myself and @pawanjay176 may take this over and I'll merge all your work into the up coming updates. 

I'll set this to blocked, until my network upgrades get merged, then we can merge this on top. :)

**Comment by b-m-f** [comment]
@AgeManning sure thing!

**Comment by AgeManning** [comment]
Closing in favour of #866

**Pull request closed without merging**